### ISSUE TITLE
Planning

### ISSUE THREAD

**Issue by codeguy** [thephpleague/uploads#1]
Just opened a wiki page to jumpstart discussions:

https://github.com/thephpleague/uploads/wiki/Planning

**Comment by codeguy** [comment]
Pushed first commit with early implementation of `src/UploadedFile.php`. For version 1.0, I'm only concerned about receiving, validating, and distributing files uploaded from a `multipart/form-data` request (such that file data is added to `$_FILES` superglobal).

You can review the code. My intentions are to have a `Uploader` class that receives a Flysystem instance. You'll add multiple `UploadedFile` instances to the `Uploader`. You can also optionally append multiple validators to each `Uploader` instance (to validate all files), or directly to a single `UploadedFile` instance (to validate only that file). Validators are applied upon upload, and if all pass, the file is passed to the Flysystem instance to be distributed to its final destination. If a validator fails, I'm not sure how I want to handle that yet.

Does all of this seem like I'm moving in the right direction?

**Comment by codeguy** [comment]
Also, code is untested. It's mainly for visual review and discussion at this point.

**Comment by brandonsavage** [comment]
I'm not sure the level of abstraction that we ought to do here in terms of dealing with the superglobal. There's no real good way to upload a file without $_FILES, but at the same time, I kinda feel dirty accessing it here.

Perhaps it might make sense to have a separate class that wraps the $_FILES superglobal, both for testing and for abstraction purposes. Maybe that's what this in fact is meant to be, and I'm just missing it, but it always felt a little dirty to have the files superglobal directly involved in the file upload class.

**Comment by codeguy** [comment]
@brandonsavage Agreed. It's a fine line to walk. This component is about file uploads, and file uploads are pretty tightly coupled to the `$_FILES` superglobal and the `is_uploaded_file()` method. How far we want to abstract away from those (and how) is something I want to figure out sooner rather than later.

**Comment by greydnls** [comment]
@codeguy I tend to agree with Brandon, mostly for testing. How would you suggest we test the createFromGlobals method?

**Comment by brandonsavage** [comment]
I think one of the best arguments for abstraction is that it lets me test my code with the actual file uploader object, and theoretically I can mock the $_FILES superglobal in some meaningful way. The counterargument would be, "well, you can just mock the file uploader object".

**Comment by codeguy** [comment]
No I agree. I liken this to Slim's `Environment` class which serves to abstract the `$_SERVER` superglobal for the same reasons. I'll tinker with a solution this weekend.

**Comment by codeguy** [comment]
Just pushed `DataSource` abstraction layer that decouples the component from the superglobal array. I'll push up some initial unit tests this weekend.

https://github.com/thephpleague/uploads/commit/d3225e9c7b30d8389913de6fb0cfb79d936c3372
